crates/ruff_python_parser/src/lexer.rs

```diff
@@ -428,6 +429,40 @@ impl<'source> Lexer<'source> {
                     self.cursor.bump();
                     value.push('\\');
                 }
+                '?' => {
+                    self.cursor.bump();
+                    let mut question_count: u8 = 1;
+                    while self.cursor.eat_char('?') {
+                        question_count += 1;
+                    }
+
+                    // The help end magic command is valid if it has 1 or 2 question marks.
+                    // If the `value` is empty i.e., we are at the start of the line,
+                    // then we can't be sure if it is a help end magic command or not.
+                    if question_count > 2
+                        || value.chars().next_back().map_or(true, is_python_whitespace)
```

Comment:
Does this mean that `??abc ??\n` is a magic command and not a magic help?

Nit:

```suggestion
                        || value.chars().last().map_or(true, is_python_whitespace)
```

Reply:
Help command (`?` / `??`) are just a specific kind of escape command[^1] so they both are the help command just that as there's a whitespace before the ending `??`, the prefixed question marks are used to determine the kind of escape command. This is how it's parsed in IPython as well because the ending question marks are parsed using regex which [isn't allowing any whitespace before the ending question marks](https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L457-L459).

[^1]: I'm going to make some updates into the terminology used across to avoid any future confusion, mainly rename `MagicCommand` to `EscapeCommand` so that `MagicKind::Magic` isn't confusing. This also how it's named in [IPython codebase](https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L329-L346).